4.x - Explicit write & close of the session

sukihub opened the pull request:
- so that any exceptions in session write go through regular error handling process
- see #13581 for details

I would like to implement some more tests:
- for the assumption, that the exception from `SessionHandlerInterface::close` propagates up correctly
- and that a call to `session_write_close` on already closed session is a noop

I'm also not sure that calling `session_register_shutdown` is a good idea, discussion in #13581 

Otherwise this should be good enough.

Commit:
4.x - Explicit write & close of the session
- so that any exceptions in session write go through regular error handling process
- see #13581 for details

markstory commented:
The test failures in travis seem relevant.

Commit:
4.x - Explicit write & close of the session
- Session::close did not have an implementation for CLI
- SessionEquals and FlashParamEquals integration test constraints cannot use Session instance anymore, since Server now closes the session (read operation would reopen it, which for CLI means that it erases $_SESSION superglobal)

sukihub commented:
Well. `Session::close` did not have any implementation for running in CLI, but `Session::start` did.

(An alternative solution would be putting an alternative implementation of `session_write_close` here https://github.com/cakephp/cakephp/blob/master/tests/TestCase/Http/server_mocks.php)

markstory commented:
> An alternative solution would be putting an alternative implementation of session_write_close here

I'm not sure that will work unless the test harness also loads the namespaced mocks (which it currently doesn't).

sukihub commented:
> I'm not sure that will work unless ...

OK, I like the solution I implemented better :)

But. PHPStan makes a good point. Sometimes, `Server::run` might get PSRs `ServerRequestInterface` and that one does not have session. Which means something else "created" the session and should probably be responsible for closing it.

Would doing something like `if ($request instanceof ServerRequest) $sesssion->close()` be a good idea?

markstory commented:
> Would doing something like if ($request instanceof ServerRequest) $sesssion->close() be a good idea?

It would.

Commit:
4.x - Explicit write & close of the session
- only executed if CakePHP ServerRequest is used (we know that it uses CakePHP Session object)
- users of different PSR request implementations do have to handle their own session manually

ADmad commented on src/TestSuite/Constraint/Session/SessionEquals.php:
```diff
@@ -65,7 +66,10 @@ public function __construct(?Session $session, string $path)
+        // Server::run calls Session::close at the end of the request.
+        // Which means, that we cannot use Session object here to access the session data.
+        // Call to Session::read will start new session (and will erase the data).
+        return Hash::get($_SESSION, $this->path) === $other;
```
Directly using `$_SESSION` here means one will never be able to mock the session class. Though I don't have an alternative suggestion.

markstory replied:
If you're using a mock, you can't use these methods anyways.

sukihub replied:
I'm not sure what did you mean by "mock the session class", but I can try to offer a solution.

`Session` currently has few `isCLI` if-s, that are true while running tests. It means that it basically uses different implementation for CLI/tests. I can imagine having completely different classes for request/cli/test sessions. And the test one could allow us to read its contents here.

markstory replied:
I was referring to the scenario where a test has mocked the session. In that case you couldn't really have used these methods before either as the read() or write() method would likely have been mocked.
